File: ruby/word-count/word_count.rb
class Phrase
  def initialize(phrase)
    @phrase = phrase
  end

  def word_count
    words = Hash.new 0
    clean_phrase = @phrase.gsub(/[^a-zA-Z\d\s']/,' ')
    clean_phrase.gsub!(/\s+'/,' ')
    clean_phrase.gsub!(/'\s+/,' ')

    clean_phrase.downcase.split.each { |word| words[word] += 1 }
    words
  end
end

File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map
  def count(sentence) do
    sentence
    |> String.downcase
    |> String.replace(~r/[!&@$%^&,:]/, "")
    |> String.split(~r/(\s+|_)/u)
    |> Enum.reduce(%{}, fn word, acc ->
      Map.put(acc, word, Map.get(acc, word, 0) + 1)
    end)
  end
end
